<a href="https://colab.research.google.com/github/cisnerosjoy/Fisicacomputacional2/blob/main/TareaPractica2EJCM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1 align="center">Física Computacional.</h1>
<h1 align="center">Semestre 2014-1</h1>

<h2>Sergio A. Alcalá Corona </h2>

<h2>Alumno:Edgar Joel Cisneros Montiel </h2>

---
### Daniel Rojo Mata
### Rodrigo Vega Vilchis
---

<h1 align="center">Programación para la física computacional</h1>

## Tarea-Práctica 2


### Programación básica para la física

---

### EJERCICIOS:

**1. Números de Catalan** Los numeros de Catalan $Cn$ son una secuencia de numeros enteros $1, 1, 2, 5, 14, 42, 132$ que juegan un papel importante en la mecanica cuantica y la teorıa de los sistemas
desordenados. (Eran fundamentales para la prueba de Eugene Wigner de la llamada
ley del semicırculo). Se definen por:

<h3 align="center">$C_{0} = 1 , C_{n+1} = {4n+2 \over n+2}C_{n}$,</h3>  

Escribe un programa que imprima en orden creciente todos los numeros de Catalan $Cn$ menores o iguales a $1\times10^{24}\$ (mil millones).

In [ ]:
from math import *
from numpy import *

In [ ]:
N=100 #cuantas veces va ser que se repetira la accion del for
C0=1 #primer termino
Cn=C0
for n in range(N):
    print(Cn) #Aqui imprime el primer termino de Cn y despues se sigue con los demas del for

    Cn=((4*n+2)/(n+2))*Cn

    #print(Cn) aqui no sale el primer termino y solo son los del for

1
1.0
2.0
5.0
14.0
42.0
132.0
429.0
1430.0
4862.0
16796.0
58786.0
208012.0
742900.0
2674440.0
9694845.0
35357670.0
129644790.0
477638700.0
1767263190.0
6564120420.0
24466267020.0
91482563640.0
343059613650.0
1289904147324.0
4861946401452.0
18367353072152.0
69533550916004.0
263747951750360.0
1002242216651368.0
3814986502092304.0
1.4544636039226908e+16
5.55340648770482e+16
2.1233613041224314e+17
8.12944042149731e+17
3.116285494907302e+18
1.1959798385860456e+19
4.5950804324621754e+19
1.7673386278700676e+20
6.804253717299761e+20
2.6221270422764934e+21
1.0113918591637903e+22
3.904442991190447e+22
1.5085347920508544e+23
5.83300119592997e+23
2.2571178540772493e+24
8.740328711533179e+24
3.386877375719107e+25
1.3132789824216945e+26
5.0955224517961745e+26
1.9782616577561618e+27
7.684785670514321e+27
2.9869166945772643e+28
1.161578714557825e+29
4.519597180279538e+29
1.75941461660882e+30
6.852456927844878e+30
2.670095285677487e+31
1.0408846028912238e+32
4.059449951275773e+32
1.583850964596121e+33


**2. La constante de Madelung**: Escribe un programa para calcular e imprimir la constante de Madelung para el cloruro
de sodio. Utiliza un valor de L tan grande como puedas, sin dejar que tu programa se
ejecutar en un tiempo razonable (un minuto o menos).

**3. Coeficientes binomiales:**

a) Utiliza esta formula para escribir una funcion llamada  $binomial(n,k)$ (o como tu
quieras) que calcule el coeficiente binomial para un $n$ y $k$ dados. Asegurate de que tu funcion devuelva la respuesta en forma de un numero entero (no flotante) y proporcione el valor correcto de 1 para el caso en que $k = 0$.

In [ ]:
from math import factorial as fac
n = int(input("Ingresa el valor de n: "))
k = int(input("ingresa el valor de k: "))
def binomial(n,k):
    if k == 1 or k == n:
        return 1

    if k > n:
        return 0

    if n>k:
        a =fac(n)
        b= fac(k)
        c= fac(n-k)
        d=a/(b*c)
        return d


int(binomial(n,k))

Ingresa el valor de n: 5
ingresa el valor de k: 3


10

b) Usando tu funcion, escribe un programa que imprima las primeras 20 lıneas del
“triangulo de Pascal’’. La nesima lınea del triangulo de Pascal contiene $n + 1$
numeros, que son los coeficientes $({n \over 0}), ({n \over 1})$, y as ́ı sucesivamente hasta $({n \over n})$.

**4. Órbitas planetarias:** En el espacio, la órbita de un cuerpo alrededor de otro (como un planeta alrededor del Sol), no necesariamente es circular. En general, toma la forma de una elipse, con el cuerpo a veces más cerca y otras más lejos. Si tenemos la distancia $\ell_1$ de máxima aproximación de un planeta al Sol (su *perihelio*), y su velocidad lineal $v_1$ en el perihelio, entonces cualquier otra propiedad de la órbita se puede calcular a partir de estas dos cantidades de la siguiente manera:

   
**a)** La segunda ley de Kepler nos dice que la distancia $\ell_2$ y la velocidad $v_2$ del planeta en su punto más distante, o *afelio*, satisfacen que $\ell_2 v_2 = \ell_1 v_1$.  Al mismo tiempo, la energía total, cinética más la gravitatoria, de un planeta con velocidad $v$ y distancia $r$ del Sol está dada por:

<h3 align="center">$E = \frac{1}{2} m v^2 - G {mM\over r}$,</h3>
        
donde $m$ es la masa del planeta, $M=1.9891\times10^{30}\,$kg es la masa del Sol y $G=6.6738\times10^{-11}\,\mathrm{m^3\,kg^{-1}\,s^{-2}}$ es la constante gravitatoria de Newton. Dado que la energía debe conservarse, demuestre que $v_2$ es la raíz más pequeña de la ecuación cuadrática:

<h3 align="center">$v_2^2 - {2GM\over v_1\ell_1} v_2 - \biggl[ v_1^2 - {2GM\over\ell_1} \biggr] = 0$.</h3>

Una vez que tenemos $v_2$ podemos calcular $\ell_2$  usando la relación $\ell_2 = \ell_1 v_1/v_2$.

**b)** Dados los valores de $v_1$, $\ell_1$, y $\ell_2$; otros parámetros de la órbita se obtienen mediante fórmulas simples que pueden derivarse de las leyes de Kepler y del hecho de que la órbita es una elipse:

<h4 align="center">Semieje mayor: $\; \; \; a = \frac{1}{2}(\ell_1+\ell_2)$,</h4>

<h4 align="center">Semieje menor: $\; \; \; b = \sqrt{\ell_1\ell_2}$,</h4>

<h4 align="center">Período orbital: $\; \; \; T = {2\pi ab\over\ell_1 v_1}$,</h4>

<h4 align="center">Excentricidad orbital: $\; \; \; e = {\ell_2-\ell_1\over\ell_2+\ell_1}$.</h4>

Escribe un programa que le pida al usuario que ingrese la distancia al Sol y la velocidad en el perihelio; para que calcule e imprima las cantidades:

$\ell_2$, $v_2$, $T$ y $e$.

In [ ]:
#Para la tierra
G=6.6738e-11
M=1.9891e30
l1=float(input("Ingresa la distancia al sol en el perihelio en metros: "))
v1=float(input("Ingresa la velocidad en el perihelio en metros sobre segundos: "))

v2 = ((2*G*M)/(v1*l1)) - v1

l2=(l1*v1)/v2


T = (pi*((l1*l2)**(1/2))*(l1 + l2))/(l1*v1)

e=(l2-l1)/(l2+l1)

print("Velocidad del planeta v2=",v2,"m/s")
print("Distancia del planeta l2=",l2,"m")
print("Periodo del planeta T=",T,"s")
print("Excentricidad del planeta e=",e)

Ingresa la distancia al sol en el perihelio en metros: 1.4710e11
Ingresa la velocidad en el perihelio en metros sobre segundos: 30287
Velocidad del planeta v2= 29305.399177261308 m/s
Distancia del planeta l2= 152027197208.65976 m
Periodo del planeta T= 31543060.207886893 s
Excentricidad del planeta e= 0.016471913134741584


**c)** Prueba tu programa haciendo que calcule las propiedades de las órbitas de la Tierra (para las cuales $\ell_1=1.4710\times10^{11}\,$m y $v_1=3.0287\times10^4\,\mathrm{m\,s^{-1}}$)  y del cometa Halley ($\ell_1=8.7830\times10^{10}\,$m y $v_1=5.4529\times10^4\,\mathrm{m\,s^{-1}}$). Entre otras cosas, deberías encontrar que el período orbital de la Tierra es de un año y el del cometa Halley es de unos 76 años.

In [ ]:
#para el cometa Halley
G=6.6738e-11
M=1.9891e30
l1=float(input("Ingresa la distancia al sol en el perihelio en metros: "))
v1=float(input("Ingresa la velocidad en el perihelio en metros sobre segundos: "))

v2 = ((2*G*M)/(v1*l1)) - v1

l2=(l1*v1)/v2


T = (pi*((l1*l2)**(1/2))*(l1 + l2))/(l1*v1)

e=(l2-l1)/(l2+l1)

print("Velocidad del planeta v2=",v2,"m/s")
print("Distancia del planeta l2=",l2,"m")
print("Periodo del planeta T=",T,"s")
print("Excentricidad del planeta e=",e)

Ingresa la distancia al sol en el perihelio en metros: 8.7830e10
Ingresa la velocidad en el perihelio en metros sobre segundos: 5.4529e4
Velocidad del planeta v2= 906.680696919153 m/s
Distancia del planeta l2= 5282214660876.42 m
Periodo del planeta T= 2399312511.8451734 s
Excentricidad del planeta e= 0.967288912645406


**5. La fórmula semiempírica de la masa (FSM)**

En física nuclear, la **fórmula de Weizsäcker** (conocida también como fórmula semiempírica) sirve para evaluar la masa y otras propiedades de un núcleo atómico; y está basada parcialmente en mediciones empíricas. En particular la fórmula se usa para calcular la \textbf{\textit{energía de enlace nuclear aproximada}} $B$, de un núcleo atómico con número atómico $Z$ y número de masa $A$:

<h3 align="center">$B = a_1 A - a_2 A^{2/3} - a_3 {Z^2\over A^{1/3}} - a_4 {(A - 2Z)^2\over A} + {a_5\over A^{1/2}}$,</h3>

donde, en unidades de millones de electrón-volts, las constantes son $a_1=15.8$, $a_2=18.3$, $a_3=0.714$, $a_4=23.2$ y          
<h6 align="center">
$a_5 =
  \left \{
    \begin{aligned}
      0 & \; \text{si} \ A \; \text{es impar,} \\
      12.0 & \; \text{si} \ A \; \text{y} \; Z \; \text{son pares (ambos),} \\
      -12.0 & \; \text{si} \ A \; \text{es par y} \; Z \; \text{impar}
    \end{aligned}
  \right.
$
</h6>

**a)** Escribe un programa que tome como entrada los valores de $A$ y $Z$, e imprima la energía de enlace $B$ para el átomo correspondiente. Usa tu programa para encontrar la energía de enlace de un átomo con $A=58$ y $Z=28$ (Hint: La respuesta correcta es alrededor de los $490\,$MeV).

In [ ]:
a1=15.8
a2=18.3
a3=0.714
a4=23.2
A=int(input("Ingresa un valor entero de A: "))
Z=int(input("Ingresa un valor entero de Z: "))

if A%2==0 and Z%2==0: #aqui formamos las condiciones que se piden de A y Z para la a5
    a5=12
elif A%2==0 and Z%2==1:
    a5=-12
elif A%2==1:
    a5=0

B=(a1*A)-(a2*(A**(2/3)))-(a3*((Z**2)/(A**(1/3))))-(a4*(((A-(2*Z))**2)/A))+(a5/(A**(1/2)))

print("LA energía de enlace nuclear aproximada es",B,"MeV")

Ingresa un valor entero de A: 58
Ingresa un valor entero de Z: 28
LA energía de enlace nuclear aproximada es 497.5620206224374 MeV


**b)** Modifica el programa del inciso anterior, para escribir una segunda versión que imprima no la energía de enlace total $B$, sino la energía de unión por nucleón, que es $B/A$.

In [ ]:
a1=15.8
a2=18.3
a3=0.714
a4=23.2
A=int(input("Ingresa un valor entero de A: "))
Z=int(input("Ingresa un valor entero de Z: "))

if A%2==0 and Z%2==0: #aqui formamos las condiciones que se piden de A y Z para la a5
    a5=12
elif A%2==0 and Z%2==1:
    a5=-12
elif A%2==1:
    a5=0

B=(a1*A)-(a2*(A**(2/3)))-(a3*((Z**2)/(A**(1/3))))-(a4*(((A-(2*Z))**2)/A))+(a5/(A**(1/2)))

print("La energía de unión por nucleón máxima es", B/A, "MeV")

Ingresa un valor entero de A: 58
Ingresa un valor entero de Z: 28
La energía de unión por nucleón máxima es 8.578655527973059 MeV


**c)** Escribe una tercera versión del programa para que tome como entrada solo un valor del número atómico $Z$ y luego pase por todos los valores de $A$ desde $A=Z$ hasta $A=3Z$, para encontrar el que tiene la mayor energía de enlace por nucleón . Este es el núcleo más estable con el número atómico dado. Haz que tu programa imprima el valor de $A$ para este núcleo más estable y el valor de la energía de enlace por nucleón.

In [ ]:
from numpy import array #importamos la biblioteca que necesitamos
a1=15.8
a2=18.3
a3=0.714
a4=23.2

Z=int(input("Ingresa el número atómico:"))
#Creamos listas vacías
M = []
N = []
#creamos un ciclo for para el valor de B/A
for n in range(1,4):
    A=n*Z
    if A%2==0 and Z%2==0: #aqui formamos las condiciones que se piden de A y Z para la a5
        a5=12
    elif A%2==0 and Z%2==1:
        a5=-12
    elif A%2==1:
        a5=0

    B=(a1*A)-(a2*(A**(2/3)))-(a3*((Z**2)/(A**(1/3))))-(a4*(((A-(2*Z))**2)/A))+(a5/(A**(1/2)))
    C=B/A

#Agregamos los valores a las listas
    M.append(C)
    N.append(A)

#Necesitamos el valor maximo
w= max(M)
x=M.index(w)
y=N[x]

print("La energía de unión por nucleón máxima es",w,"MeV, A= ",y)

Ingresa el número atómico:28
La energía de unión por nucleón máxima es 8.432655535147047 MeV, A=  56


**d)** Finalmente, escribe una cuarta versión del programa que, en lugar de tomar $Z$ como entrada, se ejecute a través de todos los valores de $Z$ de 1 a 100 e imprima el valor más estable de $A$ para cada uno. ¿A qué valor de $Z$ se produce la energía de enlace máxima por nucleón? (La respuesta correcta, en la vida real, es $Z=28$, que corresponde al Níquel).

In [ ]:
a1=15.8
a2=18.3
a3=0.714
a4=23.2
Z=int
biggerC = 0.0
biggerZ = 0.0
for m in list(range(1,101)):
    Z = m
    bigA = 0.0
    bigC = 0.0
    for n in range(1,4):
        A=n*Z
        if A%2==0 and Z%2==0:
            a5=12
        elif A%2==0 and Z%2==1:
            a5=-12
        elif A%2==1:
            a5=0
        B=(a1*A)-(a2*(A**(2/3)))-(a3*((Z**2)/(A**(1/3))))-(a4*(((A-(2*Z))**2)/A))+(a5/(A**(1/2)))
        C=B/A
        if C > bigC:
            bigC = C
            bigA = A
            bigZ = Z
        if bigC > biggerC:
            biggerC = bigC
            biggerZ = bigZ
    print(bigA)
print(biggerZ)

3
4
6
8
10
12
14
16
18
20
22
24
26
28
30
32
34
36
38
40
42
44
46
48
50
52
54
56
58
60
62
64
66
68
70
72
74
76
78
80
82
84
86
88
90
92
94
96
98
100
102
104
106
108
110
112
114
116
118
120
122
124
126
128
130
132
134
136
138
140
142
144
146
148
150
152
231
234
237
240
243
246
249
252
255
258
261
264
267
270
273
276
279
282
285
288
291
294
297
300
24
